### PR TITLE
[WIP] libgit2 integration

### PR DESCRIPTION
**Pull request by hartbit**

### CODE REVIEW

File: Sources/Git/Commit.swift [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ */
+
+import clibgit
+
+/// The commit object.
+public final class Commit {
+    typealias Handle = OpaquePointer
+
+    /// The commit struct pointer.
+    let handle: OpaquePointer
```
Comment (ddunbar) [inline Sources/Git/Commit.swift]: s/OpaquePointer/Handle/, no?

File: Sources/Git/Commit.swift [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    /// The object ID of the commit.
+    public lazy var objectID: ObjectID = {
```
Comment (ddunbar) [inline Sources/Git/Commit.swift]: I believe `lazy` is an anti pattern, it isn't safe in multithreaded code.

File: Sources/Git/Commit.swift [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        return Tree(handle: treeHandle!)
+    }
+
+    deinit {
+        git_commit_free(handle)
+    }
+}
+
+extension Commit: Hashable {
+    public static func == (lhs: Commit, rhs: Commit) -> Bool {
```
Comment (ddunbar) [inline Sources/Git/Commit.swift]: style nit: `s/== (/==/`

File: Sources/Git/Commit.swift [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    public var hashValue: Int {
+        return objectID.hashValue
+    }
+}
+
+/// An action signature (e.g. for committers, taggers, etc)
+public struct Signature {
+    /// Pointer to the git signature struct (doesn't need to be freed).
```
Comment (ddunbar) [inline Sources/Git/Commit.swift]: This seems unlikely to be completely true... presumably it only doesn't need to be freed as long as some other object it is inside doesn't need to be freed? But as written, I don't think that is guaranteed? We might need this object to hold a reference to that object to make this safe.

File: Sources/Git/Commit.swift [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    // Get commit tree.
+    public func getTree() throws -> Tree {
+        var treeHandle: Tree.Handle?
+        try wrap { git_commit_tree(&treeHandle, handle) }
+        return Tree(handle: treeHandle!)
+    }
+
+    deinit {
+        git_commit_free(handle)
+    }
```
Comment (ddunbar) [inline Sources/Git/Commit.swift]: I'd recommend putting this immediately under the init

File: Sources/Git/Credentials.swift [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    override func git_cred(
+        cred: UnsafeMutablePointer<UnsafeMutablePointer<git_cred>?>?,
+        url: UnsafePointer<Int8>?,
+        username_from_url: UnsafePointer<Int8>?
+    ) -> Int32 {
+        if let username = username {
+            return git_cred_ssh_key_from_agent(cred, username)
+        }
+        return git_cred_ssh_key_from_agent(cred, username_from_url)
```
Comment (ddunbar) [inline Sources/Git/Credentials.swift]: Can this be `username ?? username_from_url`?

File: Sources/Git/Repository.swift [hunk 7]
```diff
@@ -0,0 +1,174 @@
+import clibgit
+import Basic
+import Foundation
+
+/// Error encountered during a git operation.
+public enum GitError: Error {
+    case unknown(code: Int, klass: Int, message: String?)
```
Comment (ddunbar) [inline Sources/Git/Repository.swift]: s/klass/\`class\`/ would be more consistent I think, @aciidb0mb3r ?

File: Sources/Git/Commit.swift [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+    public func next() -> ObjectID? {
+        var oid = git_oid()
+        let ret = git_revwalk_next(&oid, walker)
+
+        // Check if iterator has reached the end.
+        if ret == GIT_ITEROVER.rawValue {
+            return nil
+        }
```
Comment (ddunbar) [inline Sources/Git/Commit.swift]: I bet we could make use of a helper function which would elevate the iterator functions return value to an optional. That would simplify this, and some places below where there is a while true loop which could become `while let value = ...`

You could also probably find a way to just make them all into a real sequence.

File: Sources/Git/Commit.swift [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+        if ret != GIT_OK.rawValue {
+            // FIXME: This is an error. How to report?
```
Comment (ddunbar) [inline Sources/Git/Commit.swift]: fatalError vs silent!

File: Sources/Git/Utils.swift [hunk 10]
```diff
@@ -0,0 +1,66 @@
+import clibgit
+
+final class Box<T> {
+    let value: T
+    init(_ value: T) {
+        self.value = value
+    }
+}
```
Comment (ddunbar) [inline Sources/Git/Utils.swift]: This might as well go into Basic, IMHO.

File: Sources/Git/Utils.swift [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+public enum BranchType {
```
Comment (ddunbar) [inline Sources/Git/Utils.swift]: Wonder if this could just be `BranchType: git_branch_t` and have raw values?

File: Sources/Git/Utils.swift [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+private var isInitialized = false
+
+@discardableResult
+func wrap(_ closure: () throws -> Int32) throws -> git_error_code {
+    if !isInitialized {
+        try _wrap { git_libgit2_init() }
+    }
+
+    return try _wrap { try closure() }
+}
```
Comment (ddunbar) [inline Sources/Git/Utils.swift]: It feels kinda annoying to me to litter code with wrap.... don't almost all operations start with some kind of repository open or such where we could centralize this?

File: Sources/Git/Commit.swift (hunk 3 above)
Comment (helje5) [reply Sources/Git/Commit.swift]: `s/== (/==(/`?

**Comment by aciidb0mb3r** [comment]
Closing due to inactivity, feel free to update and reopen.

**Pull request closed without merging**